Build failure: mcpelauncher-ui-qt

### Nixpkgs version

- 1ec9dbf3abfcd01a520a5a19a25d60c2ab28fac4 (master as of right now)

### Steps to reproduce

1. `nix build github:NixOS/nixpkgs/1ec9dbf3abfcd01a520a5a19a25d60c2ab28fac4#mcpelauncher-ui-qt -L`

### Can Hydra reproduce this build failure?

Search doesn't work for me so idk.

### Relevant log output

```console
mcpelauncher-ui-qt> /build/source/mcpelauncher-ui-qt/launcherapp.cpp: In member function 'int LauncherApp::launchProfileFile(QString, QString, bool)':
mcpelauncher-ui-qt> /build/source/mcpelauncher-ui-qt/launcherapp.cpp:85:155: error: no matching function for call to 'QChar::QChar(bool&)'
mcpelauncher-ui-qt>    85 |     if(!trialMode && m_settings.value("latest_version_id").toString() != (m_latestVersion + QChar((char)m_latestVersionCode) + QChar(m_latestVersionIsBeta))) {
mcpelauncher-ui-qt>       |                                                                                                                                                           ^
mcpelauncher-ui-qt> In file included from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:14,
mcpelauncher-ui-qt>                  from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qcoreapplication.h:8,
mcpelauncher-ui-qt>                  from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtWidgets/qapplication.h:8,
mcpelauncher-ui-qt>                  from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtWidgets/QApplication:1,
mcpelauncher-ui-qt>                  from /build/source/mcpelauncher-ui-qt/launcherapp.h:4,
mcpelauncher-ui-qt>                  from /build/source/mcpelauncher-ui-qt/launcherapp.cpp:1:
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:149:43: note: candidate: 'template<class Char, typename std::enable_if<std::is_same<Char, unsigned char>::value, bool>::type <anonymous> > constexpr QChar::QChar(Char)'
mcpelauncher-ui-qt>   149 |     QT_ASCII_CAST_WARN constexpr explicit QChar(const Char c) noexcept : ucs(c) { }
mcpelauncher-ui-qt>       |                                           ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:149:43: note:   template argument deduction/substitution failed:
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:148:66: error: no type named 'type' in 'struct std::enable_if<false, bool>'
mcpelauncher-ui-qt>   148 |     template <typename Char, if_explicit_ascii_warn_char<Char> = true>
mcpelauncher-ui-qt>       |                                                                  ^~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:143:45: note: candidate: 'template<class Char, typename std::enable_if<std::is_same<Char, char>::value, bool>::type <anonymous> > constexpr QChar::QChar(Char)'
mcpelauncher-ui-qt>   143 |     QT_ASCII_CAST_WARN constexpr Q_IMPLICIT QChar(const Char ch) noexcept : ucs(uchar(ch)) {}
mcpelauncher-ui-qt>       |                                             ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:143:45: note:   template argument deduction/substitution failed:
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:142:66: error: no type named 'type' in 'struct std::enable_if<false, bool>'
mcpelauncher-ui-qt>   142 |     template <typename Char, if_implicit_ascii_warn_char<Char> = true>
mcpelauncher-ui-qt>       |                                                                  ^~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:134:24: note: candidate: 'template<class Char, typename std::enable_if<std::disjunction<std::is_same<Char, char32_t>, std::conjunction<std::is_same<Char, wchar_t>, std::integral_constant<bool, true> >, std::is_same<Char, int>, std::is_same<Char, unsigned int> >::value, bool>::type <anonymous> > constexpr QChar::QChar(Char)'
mcpelauncher-ui-qt>   134 |     constexpr explicit QChar(const Char ch) noexcept
mcpelauncher-ui-qt>       |                        ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:134:24: note:   template argument deduction/substitution failed:
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:133:66: error: no type named 'type' in 'struct std::enable_if<false, bool>'
mcpelauncher-ui-qt>   133 |     template <typename Char, if_explicit_conversion_char<Char> = true>
mcpelauncher-ui-qt>       |                                                                  ^~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:132:26: note: candidate: 'template<class Char, typename std::enable_if<std::disjunction<std::is_same<Char, short unsigned int>, std::is_same<Char, short int>, std::is_same<Char, QChar::SpecialCharacter>, std::is_same<Char, QLatin1Char>, std::conjunction<std::is_same<Char, wchar_t>, std::negation<std::integral_constant<bool, true> > >, std::is_same<Char, char16_t> >::value, bool>::type <anonymous> > constexpr QChar::QChar(Char)'
mcpelauncher-ui-qt>   132 |     constexpr Q_IMPLICIT QChar(const Char ch) noexcept : ucs(char16_t(ch)) {}
mcpelauncher-ui-qt>       |                          ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:132:26: note:   template argument deduction/substitution failed:
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:131:66: error: no type named 'type' in 'struct std::enable_if<false, bool>'
mcpelauncher-ui-qt>   131 |     template <typename Char, if_implicit_conversion_char<Char> = true>
mcpelauncher-ui-qt>       |                                                                  ^~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:119:24: note: candidate: 'constexpr QChar::QChar(uchar, uchar)'
mcpelauncher-ui-qt>   119 |     constexpr explicit QChar(uchar c, uchar r) noexcept : ucs(char16_t((r << 8) | c)) {}
mcpelauncher-ui-qt>       |                        ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:119:24: note:   candidate expects 2 arguments, 1 provided
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:115:26: note: candidate: 'constexpr QChar::QChar()'
mcpelauncher-ui-qt>   115 |     constexpr Q_IMPLICIT QChar() noexcept : ucs(0) {}
mcpelauncher-ui-qt>       |                          ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:115:26: note:   candidate expects 0 arguments, 1 provided
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note: candidate: 'constexpr QChar::QChar(const QChar&)'
mcpelauncher-ui-qt>    48 | class QT6_ONLY(Q_CORE_EXPORT) QChar {
mcpelauncher-ui-qt>       |                               ^~~~~
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note:   no known conversion for argument 1 from 'bool' to 'const QChar&'
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note: candidate: 'constexpr QChar::QChar(QChar&&)'
mcpelauncher-ui-qt> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note:   no known conversion for argument 1 from 'bool' to 'QChar&&'
mcpelauncher-ui-qt> [ 95%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/supportedandroidabis.cpp.o
mcpelauncher-ui-qt> [ 96%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/cpuid.cpp.o
mcpelauncher-ui-qt> [ 97%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/encrypted_file_login_cache.cpp.o
mcpelauncher-ui-qt> [ 98%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/qrc_qml.cpp.o
mcpelauncher-ui-qt> [ 99%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/encryption.cpp.o
mcpelauncher-ui-qt> make[2]: *** [mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/build.make:177: mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/launcherapp.cpp.o] Error 1
mcpelauncher-ui-qt> make[2]: *** Waiting for unfinished jobs....
mcpelauncher-ui-qt> make[1]: *** [CMakeFiles/Makefile2:811: mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/all] Error 2
mcpelauncher-ui-qt> make: *** [Makefile:146: all] Error 2
error: builder for '/nix/store/pmhnl436nwapl4k2qf1i1bybadrj3knx-mcpelauncher-ui-qt-1.2.0-qt6.drv' failed with exit code 2;
       last 25 log lines:
       >   131 |     template <typename Char, if_implicit_conversion_char<Char> = true>
       >       |                                                                  ^~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:119:24: note: candidate: 'constexpr QChar::QChar(uchar, uchar)'
       >   119 |     constexpr explicit QChar(uchar c, uchar r) noexcept : ucs(char16_t((r << 8) | c)) {}
       >       |                        ^~~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:119:24: note:   candidate expects 2 arguments, 1 provided
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:115:26: note: candidate: 'constexpr QChar::QChar()'
       >   115 |     constexpr Q_IMPLICIT QChar() noexcept : ucs(0) {}
       >       |                          ^~~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:115:26: note:   candidate expects 0 arguments, 1 provided
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note: candidate: 'constexpr QChar::QChar(const QChar&)'
       >    48 | class QT6_ONLY(Q_CORE_EXPORT) QChar {
       >       |                               ^~~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note:   no known conversion for argument 1 from 'bool' to 'const QChar&'
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note: candidate: 'constexpr QChar::QChar(QChar&&)'
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qchar.h:48:31: note:   no known conversion for argument 1 from 'bool' to 'QChar&&'
       > [ 95%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/supportedandroidabis.cpp.o
       > [ 96%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/cpuid.cpp.o
       > [ 97%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/encrypted_file_login_cache.cpp.o
       > [ 98%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/qrc_qml.cpp.o
       > [ 99%] Building CXX object mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/encryption.cpp.o
       > make[2]: *** [mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/build.make:177: mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/launcherapp.cpp.o] Error 1
       > make[2]: *** Waiting for unfinished jobs....
       > make[1]: *** [CMakeFiles/Makefile2:811: mcpelauncher-ui-qt/CMakeFiles/mcpelauncher-ui-qt.dir/all] Error 2
       > make: *** [Makefile:146: all] Error 2
       For full logs, run 'nix log /nix/store/pmhnl436nwapl4k2qf1i1bybadrj3knx-mcpelauncher-ui-qt-1.2.0-qt6.drv'.
```

### System metadata

- system: `"x86_64-linux"`
 - host os: `Linux 6.12.23-xanmod1, NixOS, 25.05 (Warbler), 25.05.20250421.c11863f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/nat/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/nat/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/nat/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/nat/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf:/nix/store/pz2l9m282mx9aqyskd1pjs3rh4sgyk93-gnome-settings-daemon-47.2/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/jxs248g15qklggm9gfyddq822vkcrfg2-lix-2.91.1/share`
 - nixpkgs: `/nix/store/kcmmd6alr3lx56vkf72503h3pxgf6iv4-source`

### Notify maintainers

@Aleksanaa @morxemplum @phanirithvij

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix
+++ b/pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix
@@ -3,6 +3,7 @@
   stdenv,
   mcpelauncher-client,
   fetchFromGitHub,
+  fetchpatch,
   cmake,
   pkg-config,
   zlib,
@@ -27,6 +28,16 @@ stdenv.mkDerivation (finalAttrs: {
 
   patches = [
     ./dont_download_glfw_ui.patch
+    # Qt 6.9 no longer implicitly converts non-char types (such as booleans) to
+    # construct a QChar. This leads to a build failure with Qt 6.9. Upstream
+    # has merged a patch, but has not yet formalized it through a release, so
+    # we must fetch it manually. Remove this fetch on the next point release.
+    (fetchpatch {
+      url = "https://github.com/minecraft-linux/mcpelauncher-ui-qt/commit/0526b1fd6234d84f63b216bf0797463f41d2bea0.diff";
+      hash = "sha256-vL5iqbs50qVh4BKDxTOpCwFQWO2gLeqrVLfnpeB6Yp8=";
+      stripLen = 1;
+      extraPrefix = "mcpelauncher-ui-qt/";
+    })
   ];
 
   nativeBuildInputs = [

